synapse/replication/http/_base.py

```diff
@@ -426,6 +426,8 @@ async def _check_auth_and_handle(
             code, response = await self.response_cache.wrap(
                 txn_id, self._handle_request, request, content, **kwargs
             )
+            # Take a copy so we don't mutate things in the cache.
+            response = dict(response)
```

Comment:
I am reminded of [Brenda](https://www.youtube.com/watch?v=H6-IQAdFU3w).

I think we should seriously consider landing some version of https://github.com/matrix-org/synapse/pull/13755 and https://github.com/matrix-org/synapse/pull/14911